### PR TITLE
improve docs for `@inbounds` and `Base.@propagate_inbounds`

### PR DESCRIPTION
**Pull request by aviatesk**
Despite the widespread use of these macros, there's a lack of detailed explanation about their behaviors. In particular, it's not very clear how the latter macro can be employed. This commit aims to improve their documentation, supplementing it with examples.

### CODE REVIEW

**Review by Keno — APPROVED** [review]

**Comment by jishnub** [comment]
I fear the current examples won't be easy to understand for newcomers. Perhaps we may define a custom array, and show how bounds-checking is disabled in its `getindex`?

**Comment by Seelengrab** [comment]
I don't think newcomers should use these macros without getting familiar with what those examples mean. There's something to be said about making the examples easier for advanced users that don't know about internals like `Base.arrayref`, but that still shouldn't target newcomers.

**Comment by jishnub** [comment]
By newcomers, I meant users who might want to contribute to the various array packages, but who might not be familiar with `Base` internals. Perhaps "advanced users" is the appropriate term for that group. The gap between package users and contributors is much lower in Julia than elsewhere, so ideally the documentation should focus on new contributors to the package ecosystem.

**Comment by Seelengrab** [comment]
People who are willing to contribute are definitely advanced users, yes. I agree that a bespoke example is likely better than using `Base.arrayref`.

File: base/essentials.jl [hunk 1]
```diff
@@ -700,14 +701,43 @@ end
...
+    layer deep. For example, if an `@inbounds` block includes a call to `f()`, which in turn
+    calls `g()`, bounds checks that are syntactically within `f()` will be eliminated,
+    while ones within `g()` will not.
+    If you want to eliminates bounds checks within `g()` also,
+    you need to annotate [`Base.@propagate_inbounds`](@ref) on `f()`.
+
+# Example
+
+```julia-repl
+julia> code_typed((Vector{Any},Int)) do a, i
```
Comment (KristofferC) [inline base/essentials.jl:725]: The `do` form of `code_typed` is hard to parse for me, I think it is better to extract that part into it's own function. Also, I feel that `arrayref` has to be explicitly commented on and explained that the first argument is if it should be boundschecked.

File: base/essentials.jl [hunk 2]
```diff
@@ -700,14 +701,43 @@ end
...
-    that all accesses are in bounds. In particular, using `1:length(A)` instead of
-    `eachindex(A)` in a function like the one above is _not_ safely inbounds because
+    that all accesses are in-bounds. In particular, using `1:length(A)` instead of
+    `eachindex(A)` in a function like the one above is _not_ safely in-bounds because
     the first index of `A` may not be `1` for all user defined types that subtype
     `AbstractArray`.
+
+!!! note
+    `@inbounds` eleminates bounds checks that are syntactically within the given block,
+    as well as ones in methods that are called within the block.
```
Comment (KristofferC) [inline base/essentials.jl:714]: This is tied to inlining, no?

```
julia> g() = @boundscheck error("check from g")
g (generic function with 1 method)

julia> f() = @inbounds g()
f (generic function with 1 method)

julia> f() # no error
```

```
julia> @noinline g() = @boundscheck error("check from g")
g (generic function with 1 method)

julia> f() = @inbounds g()
f (generic function with 1 method)

julia> f()
ERROR: check from g
Stacktrace:
```

Comment (vtjnash) [inline base/essentials.jl:714]:
```suggestion
    `@inbounds` eleminates bounds checks in methods called within a given block, but not
    those that are syntactically within the given block.
```

File: base/essentials.jl (hunk 1 above)
Comment (vtjnash) [reply base/essentials.jl:725]: It is also not accurate anymore, since `arrayref` is changed to `memoryrefget` and `Expr(:boundscheck)` is no longer removed, though I am fine with the do-block form. Perhaps this should show the IR for `@boundscheck return true; return false`

File: base/essentials.jl [hunk 3]
```diff
@@ -846,13 +846,14 @@ macro boundscheck(blk)
...
-    @inbounds(blk)
+    @inbounds block
 
-Eliminates array bounds checking within expressions.
-
-In the example below the in-range check for referencing
-element `i` of array `A` is skipped to improve performance.
+Eliminates bounds checking within the block.
+This macro can be used to improve performance by informing the compiler that accesses to
+array elemements or object fields are assuredly within bounds.
```
Comment (giordano) [inline base/essentials.jl:853]:
```suggestion
array elements or object fields are assuredly within bounds.
```

File: base/essentials.jl [hunk 4]
```diff
@@ -864,14 +865,44 @@ end
...
     Using `@inbounds` may return incorrect results/crashes/corruption
     for out-of-bounds indices. The user is responsible for checking it manually.
     Only use `@inbounds` when you are certain that all accesses are in bounds (as
     undefined behavior, e.g. crashes, might occur if this assertion is violated). For
     example, using `1:length(A)` instead of `eachindex(A)` in a function like
     the one above is _not_ safely inbounds because the first index of `A` may not
     be `1` for all user defined types that subtype `AbstractArray`.
+
+!!! note
+    `@inbounds` eleminates bounds checks in methods called within a given block, but not
```
Comment (giordano) [inline base/essentials.jl:877]:
```suggestion
    `@inbounds` eliminates bounds checks in methods called within a given block, but not
```

File: base/essentials.jl [hunk 5]
```diff
@@ -864,14 +865,44 @@ end
...
     be `1` for all user defined types that subtype `AbstractArray`.
+
+!!! note
+    `@inbounds` eliminates bounds checks in methods called within a given block, but not
+    those that are syntactically within the given block.
+    However, keep in mind that the `@inbounds` context propagates only one function call
+    layer deep. For example, if an `@inbounds` block includes a call to `f()`, which in turn
+    calls `g()`, bounds checks that are syntactically within `f()` will be eliminated,
+    while ones within `g()` will not.
+    If you want to eliminates bounds checks within `g()` also,
```
Comment (barucden) [inline base/essentials.jl:883]:
```suggestion
    If you want to eliminate bounds checks within `g()` also,
```

File: base/essentials.jl
Comment (Antoinemarteau) [inline base/essentials.jl:859]:
```julia
    @boundscheck block

Annotates the expression `block` as a bounds checking block, allowing it to be elided by [`@inbounds`](@ref).
```

File: base/essentials.jl [hunk 6]
```diff
@@ -864,14 +865,44 @@ end
...
+
+julia> code_typed((Vector{Any},Int)) do a, i
+           @inbounds a[i] # The bounds check for `arrayref` is turned off.
+       end |> only
+CodeInfo(
+1 ─ %1 = Base.arrayref(false, a, i)::Any
+└──      return %1
+) => Any
+```
+
```
Comment (Antoinemarteau) [inline base/essentials.jl:963]:
```julia
See also [`@boundscheck`](@ref) for custom `getindex` and `setindex!` implementations.
```

File: base/essentials.jl
Comment (Antoinemarteau) [inline base/essentials.jl:898]:
```
    indices against its [`axes`](@ref) using [`checkbounds`](@ref). 
    Therefore, `@boundscheck` annotations should only be added to a 
    [`getindex`](@ref) or [`setindex!`](@ref) implementation after you are 
    certain its behavior is correct.
```

File: base/expr.jl [hunk 7]
```diff
@@ -876,9 +876,43 @@ macro nospecializeinfer(ex)
...
+julia> code_typed((Vector{Any},Int)) do a, i
+           # Now this `@inbounds` context propagates to `getindex`,
+           # and the bounds check for `arrayref` is turned off.
+           @inbounds call_func(getindex, a, i)
+       end |> only
+CodeInfo(
+1 ─ %1 = Base.arrayref(false, a, i)::Any
+└──      return %1
+) => Any
+```
```
Comment (Antoinemarteau) [inline base/expr.jl:1105]:
```julia

See also [`@boundscheck`](@ref) for custom `getindex` and `setindex!` implementations.
```

File: base/expr.jl [hunk 8]
```diff
@@ -876,9 +876,43 @@ macro nospecializeinfer(ex)
 end
 
 """
-    @propagate_inbounds
+    Base.@propagate_inbounds function f(args...)
+        ...
+    end
+    Base.@propagate_inbounds f(args...) = ...
+
+Tells the compiler to inline `f` while retaining the caller's `@inbounds` context.
```
Comment (Antoinemarteau) [inline base/expr.jl:1074]: Tells the compiler to inline this method of `f` while retaining the caller's `@inbounds` context.

**Review by Antoinemarteau — COMMENTED** [review]
I was [recently](https://discourse.julialang.org/t/does-propagate-inbounds-affect-one-method-or-all-methods-the-function/135461) confused by the wording "function" in `Base.@propagate_inbounds`'s docstring, so I'm adding minor comments to improve this PR.

**Comment by adienes** [comment]
I find myself constantly confused about the interaction of `@propagate_inbounds` and inlining, both "natural" and via `@inline` and think this would be very helpful to merge in some form
